proxy/router/src/lib.rs

```diff
@@ -23,7 +25,7 @@ where T: Recognize,
 /// `recognize()` method is used to determine the key for a given request. This key is
 /// used to look up a route in a cache (i.e. in `Router`), or can be passed to
 /// `bind_service` to instantiate the identified route.
-pub trait Recognize {
+pub trait Recognize: Clone {
```

Comment:
One downside perhaps not immediately obvious is that `Clone` isn't object-safe, so this prevents a `Router` from ever using trait objects. We may not need them in the proxy, but it'd probably be needed in a more generic `tower-router`.

Reply:
It doesn't fully prevent it. There are ways around it w/ additional `clone_box` `clone_arc` etc... fns + a special trait obj `Recognize` wrapper....

Reply:
Is it the same/better/worse if we move the Clone requirement onto Router (i.e. `T: Recognize + Clone`)?

Reply:
It's pretty much the same.

Here's a thought: does `Recognize::bind_service` need to be `&mut self`? It feels very similar to `NewService::new_service`, which takes `&self`. If it didn't need to be mutable, then we could just put the `Recognize` inside the `Arc`, but outside the `Mutex`.

Reply:
@seanmonstar within the scope of conduit, that works just fine